File: framework/ecommerce-framework-api/src/main/java/com/sdl/ecommerce/api/Query.java
package com.sdl.ecommerce.api;

import com.sdl.ecommerce.api.model.Category;
import com.sdl.ecommerce.api.model.FacetParameter;

import java.util.ArrayList;
import java.util.List;

/**
 * Query
 *
 * Abstract base class for query implementations.
 *
 * @author nic
 */
public abstract class Query implements Cloneable {

    private Category category;
    private List<Category> categories;
    private String searchPhrase;
    private List<FacetParameter> facets;
    private int startIndex;
    private int viewSize;
    private List<QueryFilterAttribute> filterAttributes;
    private ViewType viewType;

    // TODO: Add support for sorting
    // TODO: Should this also be an interface???

    /**
     * Set category
     * @param category
     * @return this
     */
    public Query category(Category category) {
        this.category = category;
        return this;
    }

    /**
     * Set a list of categories (OR statement)
     * @param categories
     * @return this
     */
    public Query categories(List<Category> categories) {
        this.categories = categories;
        return this;
    }

    /**
     * Set search phrase
     * @param searchPhrase
     * @return this
     */
    public Query searchPhrase(String searchPhrase) {
        this.searchPhrase = searchPhrase;
        return this;
    }

    /**
     * Set facets
     * @param facets
     * @return this
     */
    public Query facets(List<FacetParameter> facets) {
        if ( this.facets == null ) {
            this.facets = facets;
        }
        else {
            this.facets.addAll(facets);
        }
        return this;
    }

    /**
     * Add a facet
     * @param facet
     * @return this
     */
    public Query facet(FacetParameter facet) {
        if ( this.facets == null ) {
            this.facets = new ArrayList<>();
        }
        this.facets.add(facet);
        return this;
    }

    /**
     * Set start index for the result (used for pagination)
     * @param startIndex
     * @return this
     */
    public Query startIndex(int startIndex) {
        this.startIndex = startIndex;
        return this;
    }

    /**
     * Set view size for the result (used for pagination)
     * @param viewSize
     * @return this
     */
    public Query viewSize(int viewSize) {
        this.viewSize = viewSize;
        return this;
    }

    /**
     * Add a filter attribute to control what information that should be included or excluded.
     * @param filterAttribute
     * @return  this
     */
    public Query filterAttribute(QueryFilterAttribute filterAttribute) {
        if ( this.filterAttributes == null ) {
            this.filterAttributes = new ArrayList<>();
        }
        this.filterAttributes.add(filterAttribute);
        return this;
    }

    /**
     * Set view type
     * @param viewType
     * @return this
     */
    public Query viewType(ViewType viewType) {
        this.viewType = viewType;
        return this;
    }

    /**
     * @return category
     */
    public Category getCategory() {
        return category;
    }

    /**
     * @return categories
     */
    public List<Category> getCategories() {
        return categories;
    }

    /**
     * @return list of facets
     */
    public List<FacetParameter> getFacets() {
        return facets;
    }

    /**
     * @return search phrase
     */
    public String getSearchPhrase() {
        return searchPhrase;
    }

    /**
     * @return start index
     */
    public int getStartIndex() {
        return startIndex;
    }

    /**
     * @return view size
     */
    public int getViewSize() {
        return viewSize;
    }

    /**
     * @return list of filter attributes
     */
    public List<QueryFilterAttribute> getFilterAttributes() {
        return filterAttributes;
    }

    /**
     * @return view type
     */
    public ViewType getViewType() {
        return viewType;
    }

    /**
     * Clone the query
     * @return new cloned instance
     */
    public Query clone() {
        try {
            return (Query) super.clone();
        }
        catch ( CloneNotSupportedException e ) {
            throw new RuntimeException("Could not clone E-Commerce query", e);
        }
    }
}

File: dxa.net/ecommerce-framework-dotnet/SDL.ECommerce.Api/Model/Query.cs
using System.Collections.Generic;

namespace SDL.ECommerce.Api.Model
{
  
    public class Query
    {  
        public Query()
        {
            Facets = new List<FacetParameter>();
        } 
        
        public Query(Query query)
        {
            Category = query.Category;
            SearchPhrase = query.SearchPhrase;
            Facets = query.Facets;
            ViewSize = query.ViewSize;
            StartIndex = query.StartIndex;
        }
             
        public ICategory Category { get; set; }
        public string CategoryId { get; set; }
        public IList<ICategory> Categories { get; set; }
        public IList<string> CategoryIds { get; set; }
        public string SearchPhrase { get; set; }
        public IList<FacetParameter> Facets { get; set; }
        public int? ViewSize { get; set; }   
        public int? StartIndex { get; set; }
        public ViewType? ViewType { get; set; }

        // TODO: Add view type + filter attributes here

        public Query Next(IProductQueryResult queryResult)
        {
            var query = new Query(this);
            query.StartIndex = queryResult.StartIndex + queryResult.ViewSize;
            if ( query.StartIndex > queryResult.TotalCount )
            {
                return null;
            }
            return query;
        }

        public Query Previous(IProductQueryResult queryResult)
        {
            var query = new Query(this);
            query.StartIndex = queryResult.StartIndex - queryResult.ViewSize;
            if ( query.StartIndex < 0 )
            {
                return null;
            }
            return query;
        }

    }
}
